log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistry.java

```diff
@@ -53,13 +57,48 @@ public final class InternalLoggerRegistry {
             new WeakHashMap<>();
 
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
-
     private final Lock readLock = lock.readLock();
-
     private final Lock writeLock = lock.writeLock();
 
+    // ReferenceQueue to track stale WeakReferences
+    private final ReferenceQueue<Logger> staleLoggerRefs = new ReferenceQueue<>();
+
     public InternalLoggerRegistry() {}
 
+    /**
+     * Expunges stale logger references from the registry.
+     */
+    private void expungeStaleEntries() {
+        Reference<? extends Logger> loggerRef;
+        while ((loggerRef = staleLoggerRefs.poll()) != null) {
+            removeLogger(loggerRef);
+        }
```

Comment:
Would you mind changing this loop as follows, please?
```
boolean locked = false;
Reference<? extends Logger> loggerRef;
while ((loggerRef = staleLoggerRefs.poll()) != null) {
    Logger logger = loggerRef.get();
    if (logger != null) {
        if (!locked) {
            writeLock.lock();
        }
        unsafeRemoveLogger(logger);
    }
}
if (locked) {
    writeLock.unlock();
}
```
See my other comment regarding the `unsafeRemoveLogger(Logger)` method.

Reply:
Note that if `loggerRef.get()` returns `null`, we still need to remove the entry that holds the `WeakReference`.

What do you think about such an approach:

1. We `poll()` the queue until its empty.
2. If there was any reference in the queue we iterate **once** over all the entries of the map and we remove **all** the invalidated `WeakReference`s, not just those that were in the queue.

As of [my previous comment](https://github.com/apache/logging-log4j2/pull/3474#discussion_r1965597651), this seems a little bit inefficient if `loggerRef.get()` is not null, but is more efficient than iterating over all entries **once-per-reference**.

Reply:
@ppkarwasz, you're right. @Suvrat1629, please proceed with the suggestion of @ppkarwasz. Though I suggest having double-checking:

```
if (!refQueue.isEmpty) {
    lock();
    try {
        if (!refQueue.isEmpty) {
            // Clear `refQueue`
            // Clean up `loggerRefByNameByMessageFactory`
        }
    } finally {
       unlock();
    }
}
```